File: Data Structures & Algorithms/kth-largest-integer-in-a-stream/submission-1.py
class KthLargest:
    # To solve this problem, I will use a Min heap approach.
    # In the constructor, I will will store nums to the member variable then heapify it, this will take O(n) time.
    # I will store k to the member variable, so we can use it in other function in the class
    # last in the constructor, I will write a while loop to pop out the minimun values while lenght on min heap until length of min heap is not greater than k
    # For the Add function, I will push the new value to the heap first, then I will write the while loop to pop the min values until
    # length of min heap is not greater than k
    #  I will return min heap[0] at the end of function because we always pop elements out until the length of heap is equal to KthLargest
    # so min_heap[0] should be the smallest value and should be the KThLargest element.
    def __init__(self, k: int, nums: List[int]):
        self.min_heap, self.k = nums, k
        heapq.heapify(self.min_heap)
        while len(self.min_heap) > k:
            heapq.heappop(self.min_heap)

    def add(self, val: int) -> int:
        if self.min_heap and val <= self.min_heap[0]:
            return self.min_heap[0]
        heapq.heappush(self.min_heap, val)
        while len(self.min_heap) > self.k:
            heapq.heappop(self.min_heap)
        return self.min_heap[0]

    # The Time complexity for heapify is O(n)
    # The Time complexity for popping process is O(n-k * log n) for both while loops in constructor and add function
    # The Time complexity for adding new element is O(log n)
    # The Space complexity for this solution is O(k)

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number[]}
     */
    rightSideView(root) {
        // As I can see we need to check each node from each level, first we need to check the right node, if the right node does exist
        // we pick the right node, but if not, we check if the left node exist and then push the left node instead.
        // So the solve this problem, I will use a Breadth-First search approach with the queue data structure.
        // I'm going to traverse each node in level order to check which node I need to pick
        // 1. I will check if the root node does exist, if yes I will put it into the queue
        // 2. I will declare a result array for collecting the values
        // 3. I will use while loop to iterate while the length of queue is greater than 0
        // 4. I will declare level length variable to store the current queue length, this ensure the inner loop always run k times which represent to level length
        // 5. I will process the current node in the queue and shift it immediately and use the auxiliary variable to store the current node
        // 6. I will check if the current node has left and right child, if the right child does exist, I will push the right child value into the result array
        // if the right child doesn't exist I will check the left child, if it does exist, I will push the left child value into the result array instead
        // I will also push both children into the queue so we can continue check other children nodes at the next level
        // 7. After the loops finished, I will return the result array

        // Let's implement
        const res = [];
        const queue = [];
        if (root) {
            queue.push(root);
        }

        while (queue.length > 0) {
            const levelLength = queue.length;
            const subList = [];
            for (let i = 0; i < levelLength; i++) {
                const curr = queue.shift();
                subList.push(curr.val);

                if (curr.left) {
                    queue.push(curr.left);
                }

                if (curr.right) {
                    queue.push(curr.right);
                }
            }

            if (subList.length > 0) res.push(subList[subList.length - 1]);
        }

        return res;

        // So The Time complexity is O(n) because we must visit each node exactly once
        // The Space complexity is O(n) when the tree is balanced (or perfect), the queue will hold all the large amount of node at the bottom level, which contains roughly half of the total amount (~n/2)
        // The Best case is O(1) is the tree is a skewed tree, the queue will only hold 1 node at any given time becaues we always shift the processed node before pushing the child node.
    }
}
